internal/oidc/token/token_handler.go

```diff
@@ -187,6 +189,18 @@ func upstreamOIDCRefresh(ctx context.Context, session *psession.PinnipedSession,
 			WithDebugf("provider name: %q, provider type: %q", s.ProviderName, s.ProviderType))
 	}
 	if refreshedGroups != nil {
+		oldGroupsInterface := session.Fosite.Claims.Extra[oidc.DownstreamGroupsClaim].([]interface{})
+		oldGroups := []string{}
+		for _, i := range oldGroupsInterface {
+			oldGroups = append(oldGroups, i.(string))
+		}
+		added, removed := diffSortedGroups(oldGroups, refreshedGroups)
+		if added != "" {
+			warning.AddWarning(ctx, "", "groups have been added since login: "+added)
+		}
+		if removed != "" {
+			warning.AddWarning(ctx, "", "groups have been removed since login: "+removed)
+		}
```

Comment:
Lets make it clear who was added/remove from what:

```suggestion
		if len(added) > 0 {
			warning.AddWarning(ctx, "", fmt.Sprintf("User %q has been added to the following groups: %q", username, added))
		}
		if len(removed) > 0 {
			warning.AddWarning(ctx, "", fmt.Sprintf("User %q has been removed from the following groups: %q", username, removed))
		}
```

`%q` is the easiest way I know to reasonably format+escape group names.